db/types.go

```diff
@@ -0,0 +1,124 @@
+package db
+
+type DB interface {
+
+	// Get returns nil iff key doesn't exist. Panics on nil key.
+	Get([]byte) []byte
+
+	// Has checks if a key exists. Panics on nil key.
+	Has(key []byte) bool
+
+	// Set sets the key. Panics on nil key.
+	Set([]byte, []byte)
+	SetSync([]byte, []byte)
+
+	// Delete deletes the key. Panics on nil key.
+	Delete([]byte)
+	DeleteSync([]byte)
+
+	// Iterator over a domain of keys in ascending order. End is exclusive.
+	// Start must be less than end, or the Iterator is invalid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	Iterator(start, end []byte) Iterator
+
+	// Iterator over a domain of keys in descending order. End is exclusive.
+	// Start must be greater than end, or the Iterator is invalid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	ReverseIterator(start, end []byte) Iterator
+
+	// Releases the connection.
+	Close()
```

Comment:
no, we should panic on all errors because there's no way to safely handle nondeterministic failures meaningfully.  a panic trace provides more information than an error.

just restart your node after fixing the problem and sync again.

Reply:
That's a valid point, and yes it applies for our tendermint CLI. However, IMHO if we are making an SDK, for users that aren't going to necessarily make CLI apps panicking is going to trip them out. If I put this in a webapp or even a UI app, I would rather handle the error than letting my app crash.

Reply:
The SDK is specifically for building deterministic state machines.  If there's a problem with your database, it's not an error you're going to handle in your app logic. We can provide a recover function somewhere at the top of the sdk execution context that will return errors and tell them their code or devices are broken (with stack trace)